File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void sum(vector<int>candidates,int index,int n,int target,vector<int>&temp,vector<vector<int>>&ans){
        if(target==0){
            ans.push_back(temp);
            return;
        }
        if(index==n || target<0){
            return;
        }
        
        sum(candidates,index+1,n,target,temp,ans);
        temp.push_back(candidates[index]);
        sum(candidates,index,n,target-candidates[index],temp,ans);
        temp.pop_back();
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int>temp;
        vector<vector<int>>ans;
        int n=candidates.size();
        sum(candidates,0,n,target,temp,ans);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def fun(self, i, target, temp, candidates, ans):
        
        if target == 0:
            ans.append(temp.copy())
            return
        
        if i == len(candidates) or target < 0:
            return
        
        self.fun(i+1, target, temp, candidates, ans)

        temp.append(candidates[i])
        self.fun(i, target-candidates[i], temp, candidates, ans)
        temp.pop()


    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        temp = []

        self.fun(0, target, temp, candidates, ans)

        return ans